cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,225 @@
+  auto sliced_child = explode_col.get_sliced_child(stream);
+  GatherType gather_map(sliced_child.size(), stream);
+
+  // Sliced columns may require rebasing of the offsets.
+  auto offsets = explode_col.offsets_begin();
+  // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    offsets + 1, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
+  auto counting_iter = thrust::make_counting_iterator(0);
+
+  PositionType pos(sliced_child.size(), stream, mr);
+
+  // This looks like an off-by-one bug, but what is going on here is that we need to reduce each
+  // result from `lower_bound` by 1 to build the correct gather map. This can be accomplished by
+  // skipping the first entry and using the result of `lower_bound` directly.
+  thrust::transform(rmm::exec_policy(stream),
+                    counting_iter,
+                    counting_iter + gather_map.size(),
+                    gather_map.begin(),
+                    [position_array = pos.data(),
+                     offsets_minus_one,
+                     offsets,
+                     offset_size = explode_col.size()] __device__(auto idx) -> size_type {
+                      auto lb_idx =
+                        thrust::lower_bound(
+                          thrust::seq, offsets_minus_one, offsets_minus_one + offset_size, idx) -
+                        offsets_minus_one;
+                      position_array[idx] = idx - (offsets[lb_idx] - offsets[0]);
+                      return lb_idx;
+                    });
+
+  GatherType empty{0, stream};
+  return build_table(input_table,
+                     explode_column_idx,
+                     sliced_child,
+                     gather_map,
+                     empty,
+                     std::move(pos),
+                     true,
+                     stream,
+                     mr);
+}
+
+std::unique_ptr<table> explode_outer(table_view const& input_table,
+                                     size_type const explode_column_idx,
+                                     bool include_position,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view explode_col{input_table.column(explode_column_idx)};
+  auto sliced_child  = explode_col.get_sliced_child(stream);
+  auto counting_iter = thrust::make_counting_iterator(0);
+  auto offsets       = explode_col.offsets_begin();
+
+  // number of nulls or empty lists found so far in the explode column
+  GatherType null_offset(explode_col.size(), stream);
+
+  auto null_iter = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0),
+    [offsets, offsets_size = explode_col.size() - 1] __device__(int idx) {
+      return (idx > offsets_size || (offsets[idx + 1] != offsets[idx])) ? 0 : 1;
+    });
```

Comment:
I think this naming is a little more descriptive and corresponds better with the `null_or_empty_count` naming down below. 
```suggestion
  auto null_or_empty = thrust::make_transform_iterator(
    thrust::make_counting_iterator(0),
    [offsets, offsets_size = explode_col.size() - 1] __device__(int idx) {
      return (idx > offsets_size || (offsets[idx + 1] != offsets[idx])) ? 0 : 1;
    });
```